File: packages/modules/chess/rust/src/core/engine/transposition.rs
use std::sync::OnceLock;

use crate::core::types::{Board, CastlingRights, Color, Move, Square};
use crate::core::utils::file_of;

// ── Zobrist Keys ──

pub struct Zobrist {
    pub piece_keys: [[[u64; 64]; 2]; 6],
    pub side_key: u64,
    pub castling_keys: [u64; 4],
    pub ep_keys: [u64; 8],
}

fn splitmix64(seed: &mut u64) -> u64 {
    *seed = seed.wrapping_add(0x9E3779B97F4A7C15);
    let mut z = *seed;
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58476D1CE4E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D049BB133111EB);
    z ^ (z >> 31)
}

fn init_zobrist() -> Zobrist {
    let mut seed: u64 = 123456789;
    let mut next = || splitmix64(&mut seed);

    let mut piece_keys = [[[0u64; 64]; 2]; 6];
    for pt in 0..6 {
        for c in 0..2 {
            for sq in 0..64 {
                piece_keys[pt][c][sq] = next();
            }
        }
    }

    Zobrist {
        piece_keys,
        side_key: next(),
        castling_keys: [next(), next(), next(), next()],
        ep_keys: [
            next(), next(), next(), next(),
            next(), next(), next(), next(),
        ],
    }
}

fn zobrist() -> &'static Zobrist {
    static KEYS: OnceLock<Zobrist> = OnceLock::new();
    KEYS.get_or_init(init_zobrist)
}

pub fn compute_hash(
    board: &Board,
    turn: Color,
    castling_rights: CastlingRights,
    en_passant: Option<Square>,
) -> u64 {
    let keys = zobrist();
    let mut hash = 0u64;
    for sq in 0..64u8 {
        if let Some(piece) = board[sq as usize] {
            hash ^= keys.piece_keys[piece.piece_type as usize][piece.color as usize][sq as usize];
        }
    }
    if turn == Color::Black {
        hash ^= keys.side_key;
    }
    if castling_rights.K {
        hash ^= keys.castling_keys[0];
    }
    if castling_rights.Q {
        hash ^= keys.castling_keys[1];
    }
    if castling_rights.k {
        hash ^= keys.castling_keys[2];
    }
    if castling_rights.q {
        hash ^= keys.castling_keys[3];
    }
    if let Some(ep) = en_passant {
        hash ^= keys.ep_keys[file_of(ep) as usize];
    }
    hash
}

// ── Transposition Table ──

const TT_SIZE: usize = 1 << 20;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TTFlag {
    Exact,
    Alpha,
    Beta,
}

#[derive(Debug, Clone, Copy)]
pub struct TTEntry {
    pub key: u64,
    pub depth: u8,
    pub score: i32,
    pub flag: TTFlag,
    pub best_move: Option<Move>,
}

pub struct TranspositionTable {
    entries: Vec<TTEntry>,
}

impl TranspositionTable {
    pub fn new() -> Self {
        Self {
            entries: vec![
                TTEntry {
                    key: 0,
                    depth: 0,
                    score: 0,
                    flag: TTFlag::Exact,
                    best_move: None,
                };
                TT_SIZE
            ],
        }
    }

    fn index(&self, hash: u64) -> usize {
        hash as usize & (TT_SIZE - 1)
    }

    pub fn probe(&self, hash: u64) -> Option<TTEntry> {
        let entry = self.entries[self.index(hash)];
        if entry.key == hash {
            Some(entry)
        } else {
            None
        }
    }

    pub fn get_cutoff(&self, hash: u64, depth: u8, alpha: i32, beta: i32) -> Option<(i32, Option<Move>)> {
        let entry = self.entries[self.index(hash)];
        if entry.key == hash && entry.depth >= depth {
            match entry.flag {
                TTFlag::Exact => return Some((entry.score, entry.best_move)),
                TTFlag::Alpha if entry.score <= alpha => {
                    return Some((entry.score, entry.best_move))
                }
                TTFlag::Beta if entry.score >= beta => {
                    return Some((entry.score, entry.best_move))
                }
                _ => {}
            }
        }
        None
    }

    pub fn store(
        &mut self,
        hash: u64,
        depth: u8,
        score: i32,
        flag: TTFlag,
        best_move: Option<Move>,
    ) {
        let idx = self.index(hash);
        self.entries[idx] = TTEntry {
            key: hash,
            depth,
            score,
            flag,
            best_move,
        };
    }

    pub fn clear(&mut self) {
        for entry in self.entries.iter_mut() {
            entry.key = 0;
        }
    }
}

File: packages/modules/chess/go/core/engine/transposition.go
package engine

import (
	"github.com/hieudoanm/chess/core/types"
	"github.com/hieudoanm/chess/core/utils"
)

// ── Zobrist Keys ──

type zobristKeys struct {
	pieceKeys    [6][2][64]uint64
	sideKey      uint64
	castlingKeys [4]uint64
	epKeys       [8]uint64
}

var zobristCache *zobristKeys

func splitmix64(seed *uint64) uint64 {
	*seed = *seed + 0x9E3779B97F4A7C15
	z := *seed
	z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9
	z = (z ^ (z >> 27)) * 0x94D049BB133111EB
	return z ^ (z >> 31)
}

func initZobrist() *zobristKeys {
	var seed uint64 = 123456789
	next := func() uint64 { return splitmix64(&seed) }

	var keys zobristKeys
	for pt := 0; pt < 6; pt++ {
		for c := 0; c < 2; c++ {
			for sq := 0; sq < 64; sq++ {
				keys.pieceKeys[pt][c][sq] = next()
			}
		}
	}
	keys.sideKey = next()
	for i := 0; i < 4; i++ {
		keys.castlingKeys[i] = next()
	}
	for i := 0; i < 8; i++ {
		keys.epKeys[i] = next()
	}
	return &keys
}

func zobrist() *zobristKeys {
	if zobristCache == nil {
		zobristCache = initZobrist()
	}
	return zobristCache
}

const ttSize = 1 << 20

type TTFlag uint8

const (
	TTExact TTFlag = iota
	TTAlpha
	TTBeta
)

type TTEntry struct {
	key      uint64
	depth    uint8
	score    int32
	flag     TTFlag
	bestMove *types.Move
}

type TranspositionTable struct {
	entries []TTEntry
}

func NewTranspositionTable() *TranspositionTable {
	tt := &TranspositionTable{entries: make([]TTEntry, ttSize)}
	for i := range tt.entries {
		tt.entries[i] = TTEntry{bestMove: nil}
	}
	return tt
}

func (tt *TranspositionTable) index(hash uint64) int {
	return int(hash & (ttSize - 1))
}

func (tt *TranspositionTable) Probe(hash uint64) *TTEntry {
	entry := &tt.entries[tt.index(hash)]
	if entry.key == hash {
		return entry
	}
	return nil
}

func (tt *TranspositionTable) GetCutoff(hash uint64, depth uint8, alpha, beta int32) (int32, *types.Move, bool) {
	entry := tt.Probe(hash)
	if entry == nil || entry.depth < depth {
		return 0, nil, false
	}
	switch entry.flag {
	case TTExact:
		return entry.score, entry.bestMove, true
	case TTAlpha:
		if entry.score <= alpha {
			return entry.score, entry.bestMove, true
		}
	case TTBeta:
		if entry.score >= beta {
			return entry.score, entry.bestMove, true
		}
	}
	return 0, nil, false
}

func (tt *TranspositionTable) Store(hash uint64, depth uint8, score int32, flag TTFlag, bestMove *types.Move) {
	idx := tt.index(hash)
	tt.entries[idx] = TTEntry{key: hash, depth: depth, score: score, flag: flag, bestMove: bestMove}
}

func (tt *TranspositionTable) Clear() {
	for i := range tt.entries {
		tt.entries[i].key = 0
	}
}

func ComputeHash(board types.Board, turn types.Color, cr types.CastlingRights, enPassant types.Square) uint64 {
	keys := zobrist()
	var hash uint64

	ptIdx := map[types.PieceType]int{
		types.Pawn: 0, types.Knight: 1, types.Bishop: 2,
		types.Rook: 3, types.Queen: 4, types.King: 5,
	}

	for sq := 0; sq < 64; sq++ {
		if piece := board[sq]; piece != nil {
			hash ^= keys.pieceKeys[ptIdx[piece.Type]][colorIndex(piece.Color)][sq]
		}
	}

	if turn == types.Black {
		hash ^= keys.sideKey
	}
	if cr.WK {
		hash ^= keys.castlingKeys[0]
	}
	if cr.WQ {
		hash ^= keys.castlingKeys[1]
	}
	if cr.BK {
		hash ^= keys.castlingKeys[2]
	}
	if cr.BQ {
		hash ^= keys.castlingKeys[3]
	}
	if enPassant != types.NoSquare {
		hash ^= keys.epKeys[utils.FileOf(enPassant)]
	}
	return hash
}

func colorIndex(c types.Color) int {
	if c == types.White {
		return 0
	}
	return 1
}
